src/librbd/api/Snapshot.cc

```diff
@@ -222,6 +226,248 @@ int Snapshot<I>::get_id(I *ictx, const std::string& snap_name, uint64_t *snap_id
     return 0;
   }
 
+template <typename I>
+int Snapshot<I>::list(I *ictx, vector<snap_info_t>& snaps)
+{
+  ldout(ictx->cct, 20) << "snap_list " << ictx << dendl;
+
+  int r = ictx->state->refresh_if_required();
+  if (r < 0)
+    return r;
+
+  std::shared_lock l{ictx->image_lock};
+  for (map<snap_t, SnapInfo>::iterator it = ictx->snap_info.begin();
```

Comment:
Nit: might as well clean this up to use the `for(auto& it : ictx->snap_info)`-style